File: TasksList/Python/Cli/tasks.py
import sys
from datetime import datetime


def _next_id(tasks):
    if not tasks:
        return 1
    return max(t["id"] for t in tasks) + 1


def add_task(tasks, title, description):
    task = {
        "id": _next_id(tasks),
        "title": title,
        "description": description,
        "completed": False,
        "created_at": datetime.now().isoformat(),
    }
    tasks.append(task)
    return task["id"]


def list_tasks(tasks):
    if not tasks:
        print("No tasks found.")
        return
    for t in tasks:
        status = "[x]" if t["completed"] else "[ ]"
        print(f"{status} {t['id']}. {t['title']} — {t['created_at']}")


def complete_task(tasks, id):
    for t in tasks:
        if t["id"] == id:
            t["completed"] = True
            return True
    return False


def delete_task(tasks, id):
    for i, t in enumerate(tasks):
        if t["id"] == id:
            tasks.pop(i)
            return True
    return False


def main():
    tasks = []
    while True:
        print("\n=== Tasks List ===")
        print("1. Add Task")
        print("2. List Tasks")
        print("3. Complete Task")
        print("4. Delete Task")
        print("5. Exit")
        choice = input("Choose an option: ").strip()

        if choice == "1":
            title = input("Title: ").strip()
            description = input("Description: ").strip()
            add_task(tasks, title, description)
            print("Task added.")
        elif choice == "2":
            list_tasks(tasks)
        elif choice == "3":
            try:
                tid = int(input("Task ID to complete: ").strip())
                if complete_task(tasks, tid):
                    print("Task completed.")
                else:
                    print("Task not found.")
            except ValueError:
                print("Invalid ID.")
        elif choice == "4":
            try:
                tid = int(input("Task ID to delete: ").strip())
                if delete_task(tasks, tid):
                    print("Task deleted.")
                else:
                    print("Task not found.")
            except ValueError:
                print("Invalid ID.")
        elif choice == "5":
            print("Goodbye!")
            break
        else:
            print("Invalid option.")


if __name__ == "__main__":
    main()

File: TasksList/Ruby/Cli/tasks.rb
require 'time'

def next_id(tasks)
  tasks.empty? ? 1 : tasks.map { |t| t[:id] }.max + 1
end

def add_task(tasks, title, description)
  task = {
    id: next_id(tasks),
    title: title,
    description: description,
    completed: false,
    created_at: Time.now.iso8601
  }
  tasks << task
  task[:id]
end

def list_tasks(tasks)
  if tasks.empty?
    puts 'No tasks found.'
    return
  end
  tasks.each do |t|
    status = t[:completed] ? '[x]' : '[ ]'
    puts "#{status} #{t[:id]}. #{t[:title]} - #{t[:created_at]}"
  end
end

def complete_task(tasks, id)
  task = tasks.find { |t| t[:id] == id }
  task[:completed] = true if task
  !task.nil?
end

def delete_task(tasks, id)
  index = tasks.index { |t| t[:id] == id }
  tasks.delete_at(index) if index
  !index.nil?
end

def main
  tasks = []
  loop do
    puts "\n=== Tasks List ==="
    puts '1. Add Task'
    puts '2. List Tasks'
    puts '3. Complete Task'
    puts '4. Delete Task'
    puts '5. Exit'
    print 'Choose an option: '
    choice = gets.to_s.strip

    case choice
    when '1'
      print 'Title: '
      title = gets.to_s.strip
      print 'Description: '
      description = gets.to_s.strip
      add_task(tasks, title, description)
      puts 'Task added.'
    when '2'
      list_tasks(tasks)
    when '3'
      print 'Task ID to complete: '
      tid = Integer(gets.to_s.strip, exception: false)
      unless tid
        puts 'Invalid ID.'
        next
      end
      if complete_task(tasks, tid)
        puts 'Task completed.'
      else
        puts 'Task not found.'
      end
    when '4'
      print 'Task ID to delete: '
      tid = Integer(gets.to_s.strip, exception: false)
      unless tid
        puts 'Invalid ID.'
        next
      end
      if delete_task(tasks, tid)
        puts 'Task deleted.'
      else
        puts 'Task not found.'
      end
    when '5'
      puts 'Goodbye!'
      break
    else
      puts 'Invalid option.'
    end
  end
end

main if __FILE__ == $PROGRAM_NAME